sync_check/util/diff.go

```diff
@@ -0,0 +1,574 @@
+		db1:          db1,
+		db2:          db2,
+		timeField:    timeField,
+		beginTime:    beginTime,
+		endTime:      endTime,
+		splitField:   splitField,
+		dbName:       dbName,
+		chunkSize:    chunkSize,
+		sample:       sample,
+		checkThCount: checkThCount,
+	}
+}
+
+// Equal tests whether two database have same data and schema.
+func (df *Diff) Equal() (equal bool, err error) {
+	equal = true
+	tbls1, err := getTables(df.db1)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+
+	tbls2, err := getTables(df.db2)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+
+	eq := equalStrings(tbls1, tbls2)
+	if !eq {
+		log.Infof("show tables get different table. [source db tables] %v [target db tables] %v", tbls1, tbls2)
+		equal = false
+	}
+
+	for _, tblName := range tbls1 {
+		eq, err = df.EqualIndex(tblName)
+		if err != nil {
+			err = errors.Trace(err)
+			return
+		}
+		if !eq {
+			log.Infof("table have different index: %s\n", tblName)
+			equal = false
+		}
+
+		eq, err = df.EqualTable(tblName)
+		if err != nil || !eq {
+			err = errors.Trace(err)
+			equal = false
+			return
+		}
+	}
+
+	return
+}
+
+// EqualTable tests whether two database table have same data and schema.
+func (df *Diff) EqualTable(tblName string) (bool, error) {
+	eq, err := df.equalCreateTable(tblName)
+	if err != nil {
+		return eq, errors.Trace(err)
+	}
+	if !eq {
+		log.Infof("table have different schema: %s\n", tblName)
+		return eq, err
+	}
+
+	eq, err = df.equalTableData(tblName)
+	if err != nil {
+		return eq, errors.Trace(err)
+	}
+	if !eq {
+		log.Infof("table data different: %s\n", tblName)
+	}
+	return eq, err
+}
+
+// EqualIndex tests whether two database index are same.
+func (df *Diff) EqualIndex(tblName string) (bool, error) {
+	index1, err := getTableIndex(df.db1, tblName)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	defer index1.Close()
+	index2, err := getTableIndex(df.db2, tblName)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	defer index2.Close()
+
+	eq, err := equalRows(index1, index2, &showIndex{}, &showIndex{})
+	if err != nil || !eq {
+		return eq, errors.Trace(err)
+	}
+	return eq, nil
+}
+
+func (df *Diff) equalCreateTable(tblName string) (bool, error) {
+	_, err1 := getCreateTable(df.db1, tblName)
+	_, err2 := getCreateTable(df.db2, tblName)
+
+	if errors.IsNotFound(err1) && errors.IsNotFound(err2) {
+		return true, nil
+	}
+	if err1 != nil {
+		return false, errors.Trace(err1)
+	}
+	if err2 != nil {
+		return false, errors.Trace(err2)
+	}
+
+	// TODO ignore table schema currently
+	// return table1 == table2, nil
+	return true, nil
+}
+
+func (df *Diff) equalTableData(tblName string) (bool, error) {
+	dumpJobs, err := generateDumpJob(df.db1, df.dbName, tblName, df.timeField, df.beginTime, df.endTime, df.splitField, df.chunkSize, df.sample)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+
+	checkNums := len(dumpJobs) * df.sample / 100
+	checkNumArr := getRandomN(len(dumpJobs), checkNums)
```

Comment:
for example, the len(dumpJobs) is 10, and sample is 50, means only check 50% chunks is ok. 
the total job's id is [0, 1, 2, 3, 4, 5, 6, 7, 8, 9] , and getRandomN may return [9, 1 ,2 , 4, 7]